src/linear_algebra/dual.lean

```diff
@@ -206,3 +207,118 @@ def eval_equiv (h : dim K V < omega) : V ≃ₗ[K] dual K (dual K V) :=
 linear_equiv.of_bijective (eval K V) eval_ker (eval_range h)
 
 end vector_space
+
+section dual_pair
+
+open vector_space module module.dual linear_map function
+
+universes u v w
+variables {K : Type u} {V : Type v} {ι : Type w} [decidable_eq ι]
+variables [discrete_field K] [add_comm_group V] [vector_space K V]
+          [decidable_eq V] [decidable_eq (ι → V)] [decidable_eq $ dual K V]
+
+local notation `V'` := dual K V
+
+/-- `e` and `ε` have characteristic properties of a basis and its dual -/
+structure dual_pair (e : ι → V) (ε : ι → V') :=
+(eval : ∀ i j : ι, ε i (e j) = if i = j then 1 else 0)
+(total : ∀ {v : V}, (∀ i, ε i v = 0) → v = 0)
+[finite : ∀ v : V, fintype {i | ε i v ≠ 0}]
+
+end dual_pair
+
+namespace dual_pair
+
+local attribute [instance, priority 1] classical.prop_decidable
+
+open vector_space module module.dual linear_map function
+
+universes u v w
+variables {K : Type u} {V : Type v} {ι : Type w} [decidable_eq ι]
+variables [discrete_field K] [add_comm_group V] [vector_space K V]
+          [decidable_eq V] [decidable_eq (ι → V)] [decidable_eq $ dual K V]
+variables {e : ι → V} {ε : ι → dual K V} (h : dual_pair e ε)
+
+include  h
+
+/-- The coefficients of `v` on the basis `e` -/
+def coeffs (v : V) : ι →₀ K :=
+{ to_fun := λ i, ε i v,
+  support := by { haveI := h.finite v, exact {i : ι | ε i v ≠ 0}.to_finset },
+  mem_support_to_fun := by {intro i, rw set.mem_to_finset, exact iff.rfl } }
+
+omit h
+
+@[simp]
+lemma coeffs_apply (v : V) (i : ι) : h.coeffs v i = ε i v := rfl
+
+private def help_tcs : has_scalar K V := mul_action.to_has_scalar _ _
+
+local attribute [instance] help_tcs
+
+/-- linear combinations of elements of `e` -/
+def lc (e : ι → V) (l : ι →₀ K) : V := l.sum (λ (i : ι) (a : K), a • (e i))
+
+include h
+
+lemma dual_lc (l : ι →₀ K) (i : ι) : ε i (dual_pair.lc e l) = l i :=
```

Comment:
```suggestion
@[simp] lemma dual_lc (l : ι →₀ K) (i : ι) : ε i (dual_pair.lc e l) = l i :=
```

Reply:
Lean reports that `dual_lc` is an invalid simp lemma. I don't know why.

Reply:
Perhaps it is upset about the "side condition" `(h : dual_pair e ε)`? I'm not sure either.